.circleci/config.yml

```diff
@@ -376,68 +411,215 @@ jobs:
+      - git_fetch_all_tags
+      - checkout
+
+      # groth parameters are lazily generated by rust-fil-proofs during seal
+      - restore_cache:
+          key: v0-proof-params-{{ arch }}-{{ checksum "/tmp/rust-fil-proofs-checksum.txt" }}
+
+      - attach_workspace:
+          at: "."
+
+      - mkdir_test_results
+
+      - go_test:
+          sector_builder_tests: "<< parameters.sector_builder_tests >>"
+          unit: false
+          integration: true
+
+      - store_test_results:
+          path: test-results
+
+  functional_test_linux:
     docker:
       - image: circleci/golang:1.12.1
+    parallelism: 4
     working_directory: /go/src/github.com/filecoin-project/go-filecoin
-    resource_class: small
+    resource_class: xlarge
+    parameters:
+      <<: *sector_builder_tests_param
+      <<: *nightly_param
     steps:
       - run:
           name: Configure environment variables
           command: |
-            echo 'export PATH="/usr/local/go/bin:${HOME}/.cargo/bin:${PATH}:${HOME}/go/bin:${HOME}/.bin"' >> $BASH_ENV
-            echo 'export GOPATH="${HOME}/go"' >> $BASH_ENV
+            echo 'export PATH="${HOME}/.cargo/bin:${PATH}"' >> $BASH_ENV
+            echo 'export FILECOIN_PARAMETER_CACHE="${HOME}/filecoin-proof-parameters/"' >> $BASH_ENV
+            echo 'export FILECOIN_USE_PRECOMPILED_RUST_PROOFS=yes' >> $BASH_ENV
             echo 'export GO111MODULE=on' >> $BASH_ENV
       - add_ssh_keys:
           fingerprints:
             - "1e:73:c5:15:75:e0:e4:98:54:3c:2b:9e:e8:94:14:2e"
 
+      # The first checkout ensures we have the files needed to restore the cache
       - checkout
+
+      - run:
+          name: Check if any bash script sets xtrace
+          command: |
+            grep -r '^set\ \-\w*[x]\w*' . && exit 1 || exit 0
+      # Pull in all submodules (inc. rust-fil-proofs)
+      - update_submodules
+
+      # Save the Git SHA of the rust-fil-proofs submodule so that we can use it when creating a cache key
       - rust_fil_proofs_checksum
+
       - restore_cache:
           keys:
             - v5-go-deps-{{ .Branch }}-{{ arch }}-{{ checksum  "build/main.go" }}-{{ checksum "go.mod" }}-{{ checksum "/tmp/rust-fil-proofs-checksum.txt" }}
             - v5-go-deps-{{ arch }}-{{ checksum  "build/main.go" }}-{{ checksum "go.mod" }}-{{ checksum "/tmp/rust-fil-proofs-checksum.txt" }}
+
+        # The second checkout ensures we have the most recent code since the
+        # restore_cache step above can override the go-filecoin code with cached code
       - git_fetch_all_tags
       - checkout
+
+      # groth parameters are lazily generated by rust-fil-proofs during seal
+      - restore_cache:
+          key: v0-proof-params-{{ arch }}-{{ checksum "/tmp/rust-fil-proofs-checksum.txt" }}
+
+      - attach_workspace:
+          at: "."
+
+      - mkdir_test_results
+
+      - go_test:
+          sector_builder_tests: "<< parameters.sector_builder_tests >>"
```

Comment:
Since this is only for functional tests should `sector_builder_tests` be `false`?

Reply:
@frrist thanks for bringing this up!
the `sector_builder_tests` parameter is set to false by default. I've updated the config to only include the `sector_builder_tests` parameter in the functional test. it will only evaluate to `true` in the [run_sector_builder_tests workflow](https://github.com/filecoin-project/go-filecoin/blob/6d86d052ceb3aa0f236a7d14708800fcb47120b7/.circleci/config.yml#L803-L819).
this brings up another great a new level of granularity this parallelization brings - the sector build tests no longer have to run the unit and integration tests cc @laser
this is the [relevant commit](https://github.com/filecoin-project/go-filecoin/pull/2609/commits/6d86d052ceb3aa0f236a7d14708800fcb47120b7) - it also adjusts the nightly parameter slightly.

Reply:
At the top of this file could we add a comment that goes over these testing workflows? Just something high level that states when they run and gives some pointers on how to parse this file when reading?